tests-web/assets/js/tests.js

```diff
@@ -18,7 +18,38 @@ export const makeOpenTest = (map) => async (name) => {
     const networkLayer = await makeDotLayer(await network, { bounds })
     map.addLayer(networkLayer);
 
-    console.debug({ rawMapLayer, bounds, inputLayer, networkLayer, map });
+    // TODO store a reference to the layers so they can be cleaned up when wanted.
```

Comment:
Ah, we have a leak right now? I hit that sort of problem before using leaflet, because I wasn't clear from the docs about who owns the layer once it's added to a map. Working in Rust for so long has made it weird to consider keeping a pointer to a layer even after seemingly "moving" it into the map's ownership.

This shouldn't be hard to fix in any case, just need some kind of current layer bundle `var` declared outside this function, right? The scope and lifetime of a top-level var floating around in some `.js` is another thing about JS I have to get used to.

Reply:
It is only a "leak" in the sense that once we add these layers to the map, we never remove them. When we open another test case, we would want to remove the old one (because we don't expect the user to pan the map back to location of the previous test case to view it again). No actual memory is leaked though.

JS being garbage collected, we don't have to worry too much about memory management, but we do have to follow some sort of "ownership" (or at least "responsible guardian") pattern in order to have a program we can reason about.

In this case, we should look at how the leaflet API works to choose how to treat the layers. I think it would be fair to say that we "give the layer to the map" when we add it, and so we should consider that the map owns the layer, but that doesn't have any bearing on the memory -- we don't have to worry about the memory.

We can drop the reference to the layer we created, as we do here, and get it back later using `map.eachLayer(layer => ...)`. But we want to keep track of which layers are which, and call `map.removeLayer(layer)` at some point in the future. The thing to do would be to return the created layers from this function, so that the caller can decide how to keep track of them.

Then we need to make a decision about how to architect that state management, like you speculate. OOP ideas of encapsulation can be really good, but Functional Programming ideas are also very ergonomic (e.g. a top level `var` in a .js file is captured in a closure in order to be used later).

What we have here is not problematic in any way, I just wanted to hint at my upcoming ambitions.